src/twisted/web/http_headers.py

```diff
@@ -254,11 +292,11 @@ def getRawHeaders(self, name, default=None):
         encodedName = self._encodeName(name)
         values = self._rawHeaders.get(encodedName, default)
 
-        if isinstance(name, str) and values is not default:
+        if isinstance(name, str) and values is not default and values is not None:
```

Comment:
I imagine this was generating a `mypy` error because `values` is optional… and it looks like `_decodeValues(None)` would have raised an exception…?  So this changes behavior in that case… does that just never happen in practice?

Reply:
Yes, without this change, mypy errors out with:

```
src/twisted/web/http_headers.py:296:39: error: Argument 1 to "_decodeValues" of "Headers" has incompatible type "Optional[List[bytes]]"; expected "List[bytes]"  [arg-type]
                return self._decodeValues(values)
```

Reply:
I would argue that it would be more correct to put `assert values is not None`  into the `if` block. It's not a valid state.

Thinking… ok… what's actually going on here is that the return type for `self._rawHeaders.get` should be `List[AnyStr]`. The exception would be if `default` is not of that type (eg. `None`), but really that's not a correct type for `default`, which should also be `List[AnyStr]`, not `Any`.  And the return type for this method should be `List[AnyStr]`; the only reason it isn't is that we set `default`'s type to `Any`.

(Note the `Optional` in the return type is a bit nuts: `Optional` is a `Union`, which here contains another `Union`, which contains `Any`… all of that text just means: `Any`.)

I think this method's type is way too loose.  Other people should weigh in here, but I do think the correct thing to do here is to start by fixing the method type to:

```python
def getRawHeaders(self, name: AnyStr, default: List[AnyStr] = None) -> List[AnyStr]:
```

Certainly, an existing client might do something (arguably wrong) like: `rhs = h.getRawHeaders("foo", None)` but I think that this code would be better off with `rhs = h.getRawHeaders("foo", [])`, so it would be good if `mypy` actually pointed out to them that they could clean that up.

Reply:
OK, I put `assert values is not None` in the if block

Reply:
I tried change to:

```
def getRawHeaders(self, name: AnyStr, default: List[AnyStr] = None) -> List[AnyStr]:
```

but got:

```
src/twisted/web/http_headers.py:283:39: error: Argument 1 to "_decodeValues" of "Headers" has incompatible type "Union[List[bytes], List[str]]"; expected "Iterable[bytes]"
[arg-type]
                return self._decodeValues(values)
                                          ^
src/twisted/web/http_headers.py:284:16: error: Incompatible return value type (got "Union[List[bytes], List[str], None]", expected "List[str]")  [return-value]
            return values
                   ^
src/twisted/web/http_headers.py:284:16: error: Incompatible return value type (got "Optional[List[bytes]]", expected "List[bytes]")  [return-value]
            return values
                   ^
Found 3 errors in 1 file (checked 839 source files)
```

Reply:
OK, I got this to work.  I had to refactor the code a little bit but now  `getRawHeaders()` is:

```
 def getRawHeaders(
        self, name: AnyStr, default: List[AnyStr] = None
    ) -> Optional[List[AnyStr]]:
```

All the tests and mypy pass.

Reply:
@wsanchez does this look OK now?

Reply:
OK so this still has default as `Optional` (`mypy` will make it optional because it's defaulting to `None` because we haven't turned on `no_implicit_optional`, which I think we should…) and so we return an `Optional`.

I really think we shouldn't be returning `None` here and `self._rawHeaders.get(encodedName, None)` should instead read `self._rawHeaders.get(encodedName, [])`.  That would be a change in behavior.

@glyph I feel like you've expressed similar things but I'm not sure what compatibility requires or if I'm remembering you incorrectly.

Reply:
Here is @glyph's suggestion:

```python
def getRawHeaders(self, name: AnyStr, default: T = None) -> Union[List[AnyStr], T]:
```

That gives you a return type of `List[AnyStr]` (which makes me happy) unless you pass in a default value that isn't an `AnyStr` (which I think makes you a weirdo but whatever).

Reply:
I tried that and got:

```
src/twisted/web/http_headers.py:227:38: error: Name 'T' is not defined  [name-defined]
            self, name: AnyStr, default: T = None
                                         ^
src/twisted/web/http_headers.py:228:30: error: Name 'T' is not defined  [name-defined]
        ) -> Union[List[AnyStr], T]:
```

Reply:
Doesn't this take us in a circle back to:

```
def getRawHeaders(self, name: AnyStr, default: Any = None) -> Union[List[AnyStr], Any]:
```

Reply:
The typical definition for a non-specific type variable is:
```py
from typing import TypeVar
T = TypeVar('T')
```
While `T` and `Any` will both accept any type for the default argument, the return type will be different. For example if the method is called with `bytes` and `None` arguments, the signature with `T` will produce a return type of `Optional[List[bytes]]` while the signature with `Any` will produce a return type of `Any`.

Reply:
@wsanchez @glyph Can you be a bit more clear about what you want to be implemented for `getRawHeaders`?

Based on @wsanchez's feedback, I tried something like this:

```
    def getRawHeaders(
        self, name: AnyStr, default: Optional[_T] = None
    ) -> Union[List[AnyStr], Optional[_T]]:
        """
        Returns a list of headers matching the given name as the raw string
        given.

        @param name: The name of the HTTP header to get the values of.

        @param default: The value to return if no header with the given C{name}
            exists.

        @return: If the named header is present, a L{list} of its
            values.  Otherwise, C{default}.
        """
        encodedName = self._encodeName(name)
        values = self._rawHeaders.get(encodedName, [])
        if not values:
            return default

        if isinstance(name, str):
            return [v if isinstance(v, str) else v.decode("utf8") for v in values]
        return values
```

Would that work for you?

Reply:
Yeah sorry, you do need to define `T`, and `_T` is a better choice since the module is public.

You shouldn't need to make it `Optional`, though, since `_T` is unbounded; it's basically `Any`.

Reply:
The reason I made it `Optional` is because if I do this:

```
    _T = TypeVar('_T')

    def getRawHeaders(
        self, name: AnyStr, default: _T = None
    ) -> Union[List[AnyStr], _T]:
        """
        Returns a list of headers matching the given name as the raw string
        given.

        @param name: The name of the HTTP header to get the values of.

        @param default: The value to return if no header with the given C{name}
            exists.

        @return: If the named header is present, a L{list} of its
            values.  Otherwise, C{default}.
        """
        encodedName = self._encodeName(name)
        values = self._rawHeaders.get(encodedName, [])
        if not values:
            return default

        if isinstance(name, str):
            return [v if isinstance(v, str) else v.decode("utf8") for v in values]
        return values
```

Then mypy complains with:

```
src/twisted/web/http_headers.py:248:20: error: Incompatible return value type (got "Optional[_T]", expected "Union[List[str], _T]")  [return-value]
                return default
                       ^
src/twisted/web/http_headers.py:248:20: error: Incompatible return value type (got "Optional[_T]", expected "Union[List[bytes], _T]")  [return-value]
                return default
                       ^
Found 2 errors in 1 file (checked 839 source files)
```

Reply:
Default parameters and type vars interact in odd ways, and `None` weirder than most. I've filed https://github.com/python/mypy/issues/9468 because I couldn't find the existing bug about this (I swear there is one…)

For the time being probably just dealing with the `Optional` like this is the best policy.

Reply:
Looks like the mypy issue you filed is a dupe of https://github.com/python/mypy/issue/3737, https://github.com/python/mypy/issue/8708, https://github.com/python/mypy/issue/8739, https://github.com/python/mypy/issue/9179, https://github.com/python/mypy/issue/9432 :)

Reply:
OK, I have changed the signature of `getRawHeaders()` to:

```
    def getRawHeaders(
        self, name: AnyStr, default: Optional[_T] = None
    ) -> Union[List[AnyStr], Optional[_T]]:
```

Does this look OK now?

Reply:
> Looks like the mypy issue you filed is a dupe of [python/mypy#3737](https://github.com/python/mypy/issues/3737), [python/mypy#8708](https://github.com/python/mypy/issues/8708), [python/mypy#8739](https://github.com/python/mypy/issues/8739), [python/mypy#9179](https://github.com/python/mypy/issues/9179), [python/mypy#9432](https://github.com/python/mypy/issues/9432) :)

Yeah and I'm already following like 3 of these :)

Reply:
OK I see.  It's harmless, if superfluous.  Let's call it good for now.

Reply:
@wsanchez @glyph @twm thanks for the review!